src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/MarkMembersAsStatic.cs

```diff
@@ -85,10 +93,37 @@ public override void Initialize(AnalysisContext analysisContext)
                                 reportedAssociatedSymbols.Add(reportingSymbol);
                             }
 
-                            blockEndContext.ReportDiagnostic(reportingSymbol.CreateDiagnostic(Rule, reportingSymbol.Name));
+                            if (isAccessor || methodSymbol.IsExternallyVisible())
+                            {
+                                blockEndContext.ReportDiagnostic(reportingSymbol.CreateDiagnostic(Rule, reportingSymbol.Name));
+                            }
+                            else
+                            {
+                                internalCandidates.Add(methodSymbol);
+                            }
                         }
                     });
                 });
+
+                compilationContext.RegisterOperationAction(operationContext =>
+                {
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+                    if (!invocation.TargetMethod.IsExternallyVisible())
```

Comment:
From a performance point of view, it is probably a good idea to make sure the invokedInternalMethods collection doesn't become too large, and checking external visibility is good. However, lots of methods are going to be added that are never going to be in internalCandidates. Calling ShouldAnalyze is probably overkill (as that is a rather expensive method itself), but checking that the method is not static already, not abstract, virtual or override would probably not be a bad idea.